# ESAME FINE MODULO PYTHON MELISSA NORA

Analisi diffusione COVID-19 nel mondo 

Il committente richiede di avere un report su casi e vaccinazioni in diverse aree del mondo; a tal fine, richiede di utilizzare il dataset, curato da Our World in Data, all'indirizzo https://github.com/owid/covid-19-data/tree/master/public/data;  alla voce "🗂 Download our complete COVID-19 dataset" scaricare il dataset in formato CSV oppure in formato JSON. 

•Si richiede di verificare le dimensioni del dataset e le diciture presenti nell'intestazione; 

•Si chiede poi per ogni continente di trovare il numero di casi totali avvenuti in quello stesso continente; si chiede di non considerare eventuali locazioni che nel dataset non appartengono ad alcun continente; 

•Sempre riguardo i casi di COVID totali, si chiede di sviluppare una funzione che prenda in input il dataset e due nomi di continenti, e che ne confronti i seguenti relativi descrittori statistici: valori minimo e massimo, media, e percentuale rispetto al numero dei casi totali nel mondo (calcolati anche sulle locazioni senza indicazione di continente); 

•Si chiede poi di effettuare lo stesso tipo di analisi – anche in questo caso sviluppando una funzione ad hoc – per il numero di vaccinazioni totali per ogni continente; 

•Alla fine, basandosi sui calcoli fatti, il committente chiede di stilare un breve (tre o quattro righe) paragrafo testuale riassuntivo sulle statistiche di casi e vaccinazioni, che si concentri solo sulle differenze esistenti tra Europa, Sud America e Oceania.

In [21]:
import pandas as pd # importo pandas

In [35]:
file_covid = "F:/EPICODE/owid-covid-data.csv" #carico CSV dal mio PC 

•Si richiede di verificare le dimensioni del dataset e le diciture presenti nell'intestazione; 

In [66]:
struttura = pd.read_csv(file_covid, sep=',', header=0, dtype=str, low_memory=False) 

# metto il dataset in una variabile per poterlo leggere, imposto la prima riga come hearder

print(struttura) # stampa il dataset 


       iso_code continent     location        date total_cases new_cases  \
0           AFG      Asia  Afghanistan  2020-01-05         NaN       0.0   
1           AFG      Asia  Afghanistan  2020-01-06         NaN       0.0   
2           AFG      Asia  Afghanistan  2020-01-07         NaN       0.0   
3           AFG      Asia  Afghanistan  2020-01-08         NaN       0.0   
4           AFG      Asia  Afghanistan  2020-01-09         NaN       0.0   
...         ...       ...          ...         ...         ...       ...   
376980      ZWE    Africa     Zimbabwe  2024-01-31    266265.0       0.0   
376981      ZWE    Africa     Zimbabwe  2024-02-01    266265.0       0.0   
376982      ZWE    Africa     Zimbabwe  2024-02-02    266265.0       0.0   
376983      ZWE    Africa     Zimbabwe  2024-02-03    266265.0       0.0   
376984      ZWE    Africa     Zimbabwe  2024-02-04    266265.0       0.0   

       new_cases_smoothed total_deaths new_deaths new_deaths_smoothed  ...  \
0        

In [38]:
struttura.shape # il dataser ha 376985 righe e 67 colonne

(376985, 67)

In [39]:
struttura.sample (10) # mi studio 10 records causali

iso_code      continent                         location        date  \
113436       FRA         Europe                           France  2020-01-27   
354691       GBR         Europe                   United Kingdom  2023-08-29   
104015       ETH         Africa                         Ethiopia  2022-12-21   
287488       SPM  North America        Saint Pierre and Miquelon  2022-08-19   
31083        BLR         Europe                          Belarus  2023-04-18   
46866        VGB  North America           British Virgin Islands  2021-07-21   
6955         ASM        Oceania                   American Samoa  2022-09-18   
191746  OWID_LMC            NaN              Lower middle income  2021-11-23   
182893       LSO         Africa                          Lesotho  2022-03-19   
40787        BES  North America  Bonaire Sint Eustatius and Saba  2021-04-01   

       total_cases new_cases new_cases_smoothed total_deaths new_deaths  \
113436         3.0       0.0              0.429          NaN        0.0   
354691  24685994.0       0.0             1284.0     229674.0        0.0   
104015    496083.0       0.0            126.714       7572.0        0.0   
287488      3131.0       0.0              2.714          1.0        0.0   
31083     994037.0       0.0                0.0       7118.0        0.0   
46866       2024.0       0.0            125.286         12.0        0.0   
6955        8235.0      18.0              2.571         34.0        0.0   
191746  63572291.0       0.0          53768.429    1135157.0        0.0   
182893     32775.0       0.0              8.429        697.0        0.0   
40787       1247.0       0.0             43.143         10.0        0.0   

       new_deaths_smoothed  ... male_smokers handwashing_facilities  \
113436                 0.0  ...         35.6                    NaN   
354691              32.571  ...         24.7                    NaN   
104015                 0.0  ...          8.5                   7.96   
287488                 0.0  ...          NaN                    NaN   
31083                  0.0  ...         46.1                    NaN   
46866                1.286  ...          NaN                    NaN   
6955                   0.0  ...          NaN                    NaN   
191746            1580.857  ...          NaN                    NaN   
182893                 0.0  ...         53.9                  2.117   
40787                0.571  ...          NaN                    NaN   

       hospital_beds_per_thousand life_expectancy human_development_index  \
113436                       5.98           82.66                   0.901   
354691                       2.54           81.32                   0.932   
104015                        0.3            66.6                   0.485   
287488                        NaN           81.07                     NaN   
31083                        11.0           74.79                   0.823   
46866                         NaN           79.07                     NaN   
6955                          NaN           73.74                     NaN   
191746                        NaN             NaN                     NaN   
182893                        NaN           54.33                   0.527   
40787                         NaN           77.79                     NaN   

          population excess_mortality_cumulative_absolute  \
113436    67813000.0                                  NaN   
354691    67508936.0                                  NaN   
104015   123379928.0                                  NaN   
287488        5885.0                                  NaN   
31083      9534956.0                                  NaN   
46866        31332.0                                  NaN   
6955         44295.0                                  NaN   
191746  3432097300.0                                  NaN   
182893     2305826.0                                  NaN   
40787        27052.0                                  NaN   


In [46]:
colonne = list(struttura.columns) 
colonne_vert = '\n'.join(colonne) # perchè le voglio stampate in verticale
print (colonne_vert) 

iso_code
continent
location
date
total_cases
new_cases
new_cases_smoothed
total_deaths
new_deaths
new_deaths_smoothed
total_cases_per_million
new_cases_per_million
new_cases_smoothed_per_million
total_deaths_per_million
new_deaths_per_million
new_deaths_smoothed_per_million
reproduction_rate
icu_patients
icu_patients_per_million
hosp_patients
hosp_patients_per_million
weekly_icu_admissions
weekly_icu_admissions_per_million
weekly_hosp_admissions
weekly_hosp_admissions_per_million
total_tests
new_tests
total_tests_per_thousand
new_tests_per_thousand
new_tests_smoothed
new_tests_smoothed_per_thousand
positive_rate
tests_per_case
tests_units
total_vaccinations
people_vaccinated
people_fully_vaccinated
total_boosters
new_vaccinations
new_vaccinations_smoothed
total_vaccinations_per_hundred
people_vaccinated_per_hundred
people_fully_vaccinated_per_hundred
total_boosters_per_hundred
new_vaccinations_smoothed_per_million
new_people_vaccinated_smoothed
new_people_vaccinated_smoothed_per_hundre

•Si chiede poi per ogni continente di trovare il numero di casi totali avvenuti in quello stesso continente; si chiede di non considerare eventuali locazioni che nel dataset non appartengono ad alcun continente; 


In [79]:
# ho bisogno di convertire la colonna dei casi in numerico 
struttura['total_cases'] = pd.to_numeric(struttura['total_cases'], errors='coerce')

# tolgo tutti i nulli in continent
struttura_continente = struttura.dropna(subset=['continent'])

# Raggruppa i dati per continente e trova il massimo dei casi totali per ogni continente
casi_per_continente = struttura_continente.groupby('continent')['total_cases'].max().reset_index()
# mi sono presa il massimo perchè essendo continent un incrementale l'ultima data sarà già una somma di tutti i casi 

       continent  total_cases
0         Africa    4072636.0
1           Asia   99329249.0
2         Europe   38997490.0
3  North America  103436829.0
4        Oceania   11769858.0
5  South America   37519960.0


•Sempre riguardo i casi di COVID totali, si chiede di sviluppare una funzione che prenda in input il dataset e due nomi di continenti, e che ne confronti i seguenti relativi descrittori statistici: valori minimo e massimo, media, e percentuale rispetto al numero dei casi totali nel mondo (calcolati anche sulle locazioni senza indicazione di continente); 

In [133]:

'''voglio una funzione che dia la possibilità di scegliere due contenti 
in base al continente scelto nella colonna continenti di struttura_continente mi
dia in output min, max, mean e percentuale dei casi '''

# due input 
# total cases
# descrizione statistica: min, max, mean, percent tot casi nel mondo) 

def confronta_continenti(struttura, continente1, continente2):
   #lavoro prima sul continente 1 
    # Filtro il dataset per includere solo i casi relativi al continente1 specificati
    casi_continente1= struttura[struttura['continent'] == continente1]['total_cases']

    # mi calcolo i dati statistici su continente1
    minimo_continente1 = casi_continente1.min()
    massimo_continente1 = casi_continente1.max()
    media_continente1 = casi_continente1.mean()
    # uso la somma dei nuovi casi (non uso total case perchè è incrementale) come variabile 
    # ho dovuvo impostare new cases come numerico 
    struttura['new_cases'] = pd.to_numeric(struttura['new_cases'], errors='coerce')
    casi_totali_mondo = struttura['new_cases'].sum()
    # Calcolo le percentuali rispetto ai casi totali nel mondo per continente1
    casi_continente1_1= struttura[struttura['continent'] == continente1]['new_cases']
    percentuale_continente1 = (casi_continente1_1.sum() / casi_totali_mondo) * 100
    
    #ripeto tutto su continente 2 
    # Filtro il dataset per includere solo i casi relativi al continente2 specificati
    casi_continente2 = struttura[struttura['continent'] == continente2]['total_cases']
    # mi calcolo i dati statistici su continente2
    minimo_continente2 = casi_continente2.min()
    massimo_continente2 = casi_continente2.max()
    media_continente2 = casi_continente2.mean()
    
    # Calcolo le percentuali rispetto ai casi totali nel mondo per continente2
    casi_continente2_2 = struttura[struttura['continent'] == continente2]['new_cases']
    percentuale_continente2 = (casi_continente2_2.sum() / casi_totali_mondo) * 100
    print("Il min di casi in", continente1, "è", minimo_continente1)
    print("Il min di casi in", continente2, "è", minimo_continente2)
    print("Il max di casi in", continente1, "è", massimo_continente1)
    print("Il max di casi in", continente2, "è", massimo_continente2)       
    print("La media dei casi in", continente1, "è", media_continente1)
    print("La media dei casi in", continente2, "è", media_continente2)  
    print("La percentuale dei casi in", continente1, "è", percentuale_continente1)
    print("La percentuale dei casi in", continente2, "è", percentuale_continente2) 

In [134]:

confronta_continenti(struttura, 'Europe', 'South America')

Il min di casi in Europe è 1.0
Il min di casi in South America è 1.0
Il max di casi in Europe è 38997490.0
Il max di casi in South America è 37519960.0
La media dei casi in Europe è 2602675.778776635
La media dei casi in South America è 3056576.174385965
La percentuale dei casi in Europe è 7.677358236403552
La percentuale dei casi in South America è 2.0944909462884502


esempio Europa e Sud America:

Il min di casi in Europe è 1.0

Il min di casi in South America è 1.0

Il max di casi in Europe è 38997490.0

Il max di casi in South America è 37519960.0

La media dei casi in Europe è 2602675.778776635

La media dei casi in South America è 3056576.174385965

La percentuale dei casi in Europe è 7.7%

La percentuale dei casi in South America è 2.1%

•Si chiede poi di effettuare lo stesso tipo di analisi – anche in questo caso sviluppando una funzione ad hoc – per il numero di vaccinazioni totali per ogni continente; 

In [123]:
#basandomi sulla funzione di prima volevo che si ptoessero inserire più di due continenti selezionati 

def confronta_vaccinazioni_continenti2(struttura, *continenti):
    for continente in continenti:
        struttura['total_vaccinations'] = pd.to_numeric(struttura['total_vaccinations'], errors='coerce')
    
        vaccinazioni_continente = struttura[struttura['continent'] == continente]['total_vaccinations']

        min_vax_continente = vaccinazioni_continente.min()
        max_vax_continente = vaccinazioni_continente.max()
        mean_vax_continente = vaccinazioni_continente.mean()
        vax_totali_mondo = struttura['total_vaccinations'].sum()
        percentuale_vax_continente = (vaccinazioni_continente.sum() / vax_totali_mondo) * 100
    

        print("Il min dei vaccini in", continente, "è", min_vax_continente)
        print("Il max dei vaccini in", continente, "è", max_vax_continente)     
        print("La media dei vaccini in", continente, "è", mean_vax_continente)
        print("La percentuale dei vaccini in", continente, "è", percentuale_vax_continente)


In [124]:
# prova per vedere se la funzione funzioni
confronta_vaccinazioni_continenti2 (struttura, 'Europe', 'South America', 'Asia')

Il min dei vaccini in Europe è 0.0
Il max dei vaccini in Europe è 192221468.0
La media dei vaccini in Europe è 27275774.196870532
La percentuale dei vaccini in Europe è 1.782823009422204
Il min dei vaccini in South America è 0.0
Il max dei vaccini in South America è 486436436.0
La media dei vaccini in South America è 65958353.543699816
La percentuale dei vaccini in South America è 1.0501583501770269
Il min dei vaccini in Asia è 0.0
Il max dei vaccini in Asia è 3491077000.0
La media dei vaccini in Asia è 230728539.32374448
La percentuale dei vaccini in Asia è 10.628684534264309


•Alla fine, basandosi sui calcoli fatti, il committente chiede di stilare un breve (tre o quattro righe) paragrafo testuale riassuntivo sulle statistiche di casi e vaccinazioni, che si concentri solo sulle differenze esistenti tra Europa, Sud America e Oceania.

In [125]:
confronta_vaccinazioni_continenti2 (struttura, 'Europe', 'South America', 'Oceania')

Il min dei vaccini in Europe è 0.0
Il max dei vaccini in Europe è 192221468.0
La media dei vaccini in Europe è 27275774.196870532
La percentuale dei vaccini in Europe è 1.782823009422204
Il min dei vaccini in South America è 0.0
Il max dei vaccini in South America è 486436436.0
La media dei vaccini in South America è 65958353.543699816
La percentuale dei vaccini in South America è 1.0501583501770269
Il min dei vaccini in Oceania è 0.0
Il max dei vaccini in Oceania è 69306345.0
La media dei vaccini in Oceania è 12617372.973786408
La percentuale dei vaccini in Oceania è 0.0639019785567765


In [135]:
confronta_continenti(struttura, 'Europe', 'Oceania')

Il min di casi in Europe è 1.0
Il min di casi in Oceania è 1.0
Il max di casi in Europe è 38997490.0
Il max di casi in Oceania è 11769858.0
La media dei casi in Europe è 2602675.778776635
La media dei casi in Oceania è 355420.45812
La percentuale dei casi in Europe è 7.677358236403552
La percentuale dei casi in Oceania è 0.4490935565090396


# CASI

EUROPA

Il min di casi in Europe è 1.0

Il max di casi in Europe è 38997490.0

La media dei casi in Europe è 2602675.778776635

La percentuale dei casi in Europe è 8%


SUD AMERICA

Il min di casi in South America è 1.0

Il max di casi in South America è 37519960.0

La media dei casi in South America è 3056576.174385965

La percentuale dei casi in South America è 2%


OCEANIA

Il min di casi in Oceania è 1.0

Il max di casi in Oceania è 11769858.0

La media dei casi in Oceania è 355420.45812

La percentuale dei casi in Oceania è 0,4%


# VACCINI

EUROPA


Il min dei vaccini in Europe è 0.0

Il max dei vaccini in Europe è 192221468.0

La media dei vaccini in Europe è 27275774.196870532

La percentuale dei vaccini in Europe è 1.782823009422204


SUD AMERICA

Il min dei vaccini in South America è 0.0


Il max dei vaccini in South America è 486436436.0


La media dei vaccini in South America è 65958353.543699816


La percentuale dei vaccini in South America è 1.0501583501770269

OCEANIA


Il min dei vaccini in Oceania è 0.0


Il max dei vaccini in Oceania è 69306345.0


La media dei vaccini in Oceania è 12617372.973786408


La percentuale dei vaccini in Oceania è 0.0639019785567765

I continenti che hanno avuto più casi sono l'Europa e il Sud America, mentre l'Oceania ha avuto molti meno casi ma questo andrebbe confrontanto con la popolazione di ogni continente e vedere la percentuale rispetto la popolazione totale. Prestando attenzione alla percentuale di casi comparati con i casi mondiali, l'Europa è stata più colpita rispetto al Sud America. Per quanto riguarda le vaccinazioni lo stato che ha somministrato più vaccini tra questi tre continenti è il Sud America. Anche se controllanto le percentuali di tutti i vaccini effettuati nel mondo vediamo che l'Europa ha effetturato più vaccini, a seguire l'America Latina e in fine l'Oceania. Per valutare questi numeri si dovrebbe prendere in considerazione il numero della popolazione, l'Oceania ad esempio ha una popolazione nettamente inferiore all'Europa e al Sud America.